components/ai_chat/browser/ai_chat_tab_helper.cc

```diff
@@ -690,6 +695,35 @@ void AIChatTabHelper::MakeAPIRequestWithConversationHistoryUpdate(
                             std::move(data_received_callback));
 }
 
+void AIChatTabHelper::RetryAPIRequest() {
+  SetAPIError(mojom::APIError::None);
+  DCHECK(!chat_history_.empty());
+
+  // We're using a reverse iterator here to find the latest human turn
+  for (std::vector<ConversationTurn>::reverse_iterator rit =
+           chat_history_.rbegin();
+       rit != chat_history_.rend(); ++rit) {
+    if ((*rit).character_type == CharacterType::HUMAN) {
+      auto turn = *std::make_move_iterator(rit);
+
+      auto human_turn_iter = rit.base() - 1;
+      auto assistant_turn_iter =
+          rit.base();  // Ensure if this iter is valid before using it, as there
+                       // might be no element here
+
+      if (assistant_turn_iter != chat_history_.end() &&
+          assistant_turn_iter->character_type == CharacterType::ASSISTANT) {
```

Comment:
>  If there was an entry that we skipped over, then it would have neccessarily been Assistant

The checks are there to be extra safe especially after the BuildLlama2 prompt crash where the assumption was two entries always exist in the history vector. I agree just removing the last entry would be simple but it would make sense after we add a unittest. Also, there's a refactor on the way so there's going to be many moving parts which means potential to break.

Reply:
> If there was an entry that we skipped over, then it would have neccessarily been Assistant

This wasn't the case when the input wasn't disabled. I could spam the history with two messages back to back, and now the UI is in an error state. This was the reason for the check.

The checks are there to provide extra safety, especially after the BuildLlama2 prompt crash, where the assumption was that two entries always existed in the history vector. There are times when I found myself in an awkward state, particularly when testing with slow networks in coffee shops.

I agree that just removing the last entry would be simple, but it would make more sense after we add a unit test. Also, there's a refactor on the way, so there will be many moving parts, which increases the potential to break things.

Reply:
It still doesn't make sense to me - if there were 2 Human entries in a row, this loop would stop at the most recent one. There would be no next message. There doesn't seem to be a possibility, given the previous line of code, that you could skip over a Human entry.

Reply:
Just to reiterate, I'm **not** suggesting simply removing the last entry, I suggested removing _all_ the entries from the found Human entry to the end of the list (i.e. either only Human, or Human + 1 or more most recent Assistant entries):
```
chat_history_.erase(human_turn_iter, chat_history_.end())
```